### PR TITLE
methods-defining_methods-hard-q7

### PR DESCRIPTION
**Pull request by ketphan02**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [x] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/methods/defining_methods/hard/q7/CurrentYearMethod.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package methods.defining_methods.hard.q7;
+
+/**
+ * [J] Write a Java Program that asks users for the number of days in the current year. Then write a
+ * method that accepts that value and returns a boolean value if it is a leap year or not (true if it
+ * is a leap year, false if not). Let’s call the method “leapChecker”.
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q7/CurrentYearMethod.java:6]: Change question text to: 
`Write a Java Program that asks users for the number of days in the current year. Then write a method called leapChecker that accepts that value as an integer and returns a boolean value if it is a leap year or not (true if it is, false if not). The year is a leap year if it has 366 days. Lastly, if the method returns true, print "This year is a leap year", if it is false, print "This year isn't a leap year". `

File: src/methods/defining_methods/hard/q7/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    };
+  }
+
+  public void runMain() {
+    CurrentYearMethod.main(new String[0]);
+  }
+
+  static Stream<Arguments> possibleNumDaysForMethod() {
+    return Stream.of(
+        Arguments.of(366, true),
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q7/MainTest.java:38]: For this question, we should also specify what happens if an invalid number is entered (should probably just return false)

Comment (Hedgemon4) [reply src/methods/defining_methods/hard/q7/MainTest.java:38]: Then we also need to add test cases for this

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/methods/defining_methods/hard/q7/MainTest.java (hunk 2 above)
Comment (SamamthaH) [reply src/methods/defining_methods/hard/q7/MainTest.java:38]: I don't know, it's only to check if the year is a leap year, so any invalid input would still be false

Comment (Hedgemon4) [reply src/methods/defining_methods/hard/q7/MainTest.java]: Fair enough

File: src/methods/defining_methods/hard/q7/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    static Stream<Arguments> possibleNumDaysForMethod() {
+        return Stream.of(
+                Arguments.of(366, true),
+                Arguments.of(365, false),
+                Arguments.of(34, false),
+                Arguments.of(367, false));
+    }
+
+    static Stream<Arguments> possibleNumDaysForOutput() {
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q7/MainTest.java:45]: We should test negatives though because we did not day to assume valid input

Comment (Hedgemon4) [reply src/methods/defining_methods/hard/q7/MainTest.java:45]: We should also probably have a few more test cases (just in case).

File: src/methods/defining_methods/hard/q7/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    @ParameterizedTest
+    @MethodSource("possibleNumDaysForMethod")
+    void correctLeapCheckerMethod(int numDays, boolean isLeapYear) throws Throwable {
+        Object[][] arguments = {
+                {numDays, int.class}
+        };
+        MethodTest m = new MethodTest(CurrentYearMethod.class, "leapChecker", arguments);
+        Object output = m.callMethod();
+        String errorMessage = "Your leapChecker() method does not return the correct result";
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q7/MainTest.java:61]: Change the error message to be grammatically correct.

File: src/methods/defining_methods/hard/q7/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        CustomAssertions._assertEquals(isLeapYear, output, errorMessage);
+    }
+
+    @ParameterizedTest
+    @MethodSource("possibleNumDaysForOutput")
+    void printsCorrectLeapYearMessage(int numDays, String isLeapYearStr) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not print the correct message for this year";
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q7/MainTest.java:68]: Change error message to be grammatically correct

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/methods/defining_methods/hard/q7/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        String errorMessage = "Your leapChecker() method does not return the correct result.";
+        CustomAssertions._assertEquals(isLeapYear, output, errorMessage);
+    }
+
+    @ParameterizedTest
+    @MethodSource("possibleNumDaysForOutput")
+    void printsCorrectLeapYearMessage(int numDays, String isLeapYearStr) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not print the correct message for this year.";
+        runWithInput(numDays + "", new Clause[]{
+                new StringLiteral(isLeapYearStr),
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q7/MainTest.java:78]: Can we add an optional to the end of the placeholder if they print an extra space?

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
See my discord message

**Review by Hedgemon4 — APPROVED** [review]
